_published/0016-wharf-api-endpoints-cleanup.md

```diff
@@ -0,0 +1,888 @@
+title: "RFC-0016: wharf-api-endpoints-cleanup" # Update this with PR number/ID and feature name. Use leading zeros
+rfc_author_username: jilleJr
+rfc_author_name: Kalle Jillheden # Or same as username, if you wish
+
+# Leave these. Collaborator changes this before merging
+impl_issue_id: 0
+impl_issue_repo: iver-wharf/wharf-api
+last_modified_date: YYYY-MM-DD
+---
+
+# {{page.title}}
+
+- RFC PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pulls/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author: {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Restructuring the wharf-api by changing the paths of the endpoints, adding and
+renaming path parameters, and changing the request and response models around.
+
+While we will keep backward compatibility for at least one major version, we
+will be changing so much that this needs a full major version bump.
+
+This is not the full list of changes to be made for v5. Instead, it's
+a list of changes that will require us moving to it. Changes not listed in
+this RFC may be included in the version bump as well.
+
+## Motivation
+
+We are lacking consistency in our API, and there are some questionable
+endpoints in there. Most annoying issues are:
+
+- Some endpoints are plural, some are singular.
+  Ex: `POST /project` vs `GET /projects`.
+
+- Some are misplaced and rely on request body instead of path/query
+  parameters. Ex: `PUT /branches` instead of `PUT /project/{projectId}/branch`.
+
+- Path parameters are all lowercased, while all other query parameters are
+  camelCased.
+
+- We use the database models in our request and response specs. This is a
+  security weakness as it induces a higher risk of exposing too much data to
+  the user. With separate models we would also make the Swagger documentation
+  easier to understand, where we could elide irrelevant specifications such as
+  allowing to specify a project ID when creating a new project.
+
+## Explanation
+
+Moving from wharf-api v4 to v5, all endpoints are now in singular format.
+Some have also been moved to a different subpath.
+
+### RESTful API design
+
+Starting with v5, all endpoints now follow a more strict RESTful approach.
+
+Major changes:
+
+- POST endpoints no longer acts as "add or update", but instead solely as
+  "add".
+
+- PUT endpoints no longer acts as "add or update", but instead solely as
+  "update".
+
+- POST endpoints for searching have been removed, please refer to the GET
+  endpoints using query parameters instead of an HTTP request body to query.
```

Comment:
I would try to not be afraid of making exceptions to this rule if you need to allow for complex queries that become a pain to represent with query parameters. There is a reason Elasticsearch queries are POST's.

Reply:
The query objects in elasticsearch are super useful, but I don't want to see something similar in Wharf because I think that would make it uncessecarily complex. The existing solution is kind of a hack IMO where it does a `.Where` clause in GORM on the entire JSON request body, which is what I wanted to avoid with this new rule.

But yea, if we need a more complex solution in the future where we want to allow fine-grained queries, then it would definetly be unwise to constrain ourselves with this rule.

Reply:
I've added a note about this in the 'Alternatives' section